### PR TITLE
feat(interchain-indexer): report `catchup_complete` instead of `catchup_scan_complete`

### PR DESCRIPTION
**Pull request by EvgenKor**
# Report `catchup_complete` instead of `catchup_scan_complete`

## Summary

- `GET /api/v1/status/indexing` renames `catchup_scan_complete` to `catchup_complete` and widens what the flag means: it is now `catchup_progress_percent == 100.0 && failed_blocks == 0`, so it answers *completeness* rather than *scan coverage*.
- The old flag mirrored `CatchupProgress::scan_complete` ("the two catch-up cursors met"), so it reported a finished catch-up for a pair with unresolved holes in `indexer_failures` — exactly the misreading the "A Checkpoint Certifies Scanning, Not Correctness" gotcha exists to warn about. Reading the two records together is now the endpoint's job, not the caller's.
- The flag is derivable from two fields the same response already carries, so a client can reproduce it without trusting a hidden input.
- `scan_complete` itself is untouched: it remains the internal scan-level notion used by the indexers, the cursor tests, and ADR-007's reasoning.

## Implementation Plan

- `interchain-indexer-logic/src/indexer/progress.rs`: add `CatchupProgress::catchup_complete(failed_blocks) -> bool`. The failure-ledger aggregate is a parameter, not a lookup, so the module stays pure and DB-free. Module docs updated — `failed_blocks` is no longer described as "the only completeness signal in the payload", it is now the only input that can turn a finished scan back into an unfinished catch-up.
- `interchain-indexer-proto/proto/v1/status.proto`: field 7 renamed to `catchup_complete` (tag and `bool` type unchanged), with the rule stated in a comment. `swagger/v1/interchain-indexer.swagger.yaml` updated to match.
- `interchain-indexer-server/src/services/status.rs`: `collect_indexing_progress` now fills the field with `progress.catchup_complete(failed_blocks)` instead of `progress.scan_complete`. `failed_blocks` was already computed one line above from the same join, so no extra query.
- Docs: `README.md`, `.memory-bank/gotchas.md`, `.memory-bank/exploration-map.md`, `.memory-bank/research/message-lifecycle.md`, `.memory-bank/research/indexing-gaps-retries-and-checkpoint-safety.md`.

### Why the predicate is phrased over the percentage, not over `scan_complete`

Besides payload self-consistency, the two phrasings disagree in exactly one case, and the percentage gives the better answer there: a floor configured above the chain head (`realtime_cursor < start_block`) makes `total = 0`, which leaves `scan_complete` vacuously `true` while the reported percentage is `0.0`. Answering `catchup_complete: false` reads that misconfiguration as "not done" rather than shipping a payload that claims 100% completion at 0% progress.

The `== 100.0` float comparison is exact, not approximate: `progress_percent` is `100.0 * scanned / total` with both operands converted from the same `u64`, so it lands on exactly `100.0` whenever `scanned == total` as long as `100 * total` is exactly representable — true below `2^53 / 100 ≈ 9e13` blocks. A parametrized test pins this across chain-head magnitudes up to 1e12 blocks.

## API Changes

- **Breaking.** `ChainIndexingProgress.catchup_scan_complete` → `catchup_complete` in `GET /api/v1/status/indexing` (and the `StatusService.GetIndexingProgress` gRPC method). JSON consumers reading the old key now get `undefined`.
- **Behavior change, not just a rename.** The same pair can now report `false` where it previously reported `true`: 100% scanned with `failed_blocks > 0` is no longer a complete catch-up. Any dashboard or readiness gate keyed off this field should be re-checked.
- Proto wire compatibility is preserved: field number `7` and the `bool` type are unchanged, so binary gRPC clients decode into the same slot. Only the field *name* — and therefore the JSON key — changes.

## ENV / Config Changes

- None.

## Database / Migration Impact

- None. No schema, query, or write-path change; `indexer_failure_totals` was already being read by this handler.

## Testing

- `just check` — `cargo check` and `cargo clippy --all --all-targets --all-features -- -D warnings` clean.
- `just format` — no diff beyond the change itself.
- `just test-with-db` — full suite green, 0 failures (539 logic + 146 server unit tests, plus every integration binary).
- New unit tests in `progress.rs`: both inputs required; blocks remaining ⇒ incomplete; no checkpoint row ⇒ incomplete; floor above chain head ⇒ incomplete; and the parametrized exactness test above.
- New DB test `collect_indexing_progress_catchup_complete_requires_a_clean_failure_ledger`: two pairs with identical cursors, one with an open hole — proves the failure aggregate actually reaches the flag through the join, which the pure unit tests cannot.
- Updated assertions in `services/status.rs` and `tests/indexing_progress.rs` (the HTTP contract test) to the new key.

## Follow-ups

- Consumers of `catchup_scan_complete` (frontend, dashboards, readiness probes) need the key rename plus a look at the semantic change. `@blockscout/interchain-indexer-types` picks it up on the next `npm run build` — `types/dist` is generated, not checked in.
- The swagger YAML is a checked-in generated artifact and was edited by hand to match what `protoc-gen-openapiv2` emits for the new comment; worth a glance if the reviewer regenerates it.
- `catchup_complete: true` inherits the ledger's known limitation unchanged: a failure a handler swallows never becomes a row, so the flag still is not proof that every block was indexed. Documented in the gotcha and in ADR-005's research note; `interchain_indexer_oldest_open_hole_age_seconds` remains the operational alert.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a clear catch-up completion status that requires 100% scan progress and no failed blocks.
  * Renamed the indexing status field to `catchup_complete` across API responses and documentation.

* **Bug Fixes**
  * Prevented catch-up from being reported complete when unresolved failed ranges remain.

* **Documentation**
  * Clarified the distinction between scan progress and actual indexing completeness.

* **Tests**
  * Added coverage for failed ranges, complete scans, missing checkpoints, and progress edge cases.

### CODE REVIEW

File: interchain-indexer/.memory-bank/research/message-lifecycle.md [hunk 1]
```diff
@@ -169,7 +169,10 @@ Consequences worth internalizing:
 - **Catchup completion is "the two catchup cursors met"**, i.e. the unscanned
   interval is empty (`catchup_max_cursor < catchup_min_cursor`). It is not "the
   frontier reached `started_at_block`" — that formulation only happens to work
-  while `catchup_min_cursor` sits at `started_at_block`.
+  while `catchup_min_cursor` sits at `started_at_block`. This is completion of
+  the *scan*, and it is not the same thing as the endpoint's
+  `catchup_complete`, which additionally requires an empty failure ledger
+  (`failed_blocks == 0`) — see `indexer/progress.rs`.
```
[the remark this answers is not in the document]
Comment (EvgenKor) [reply interchain-indexer/.memory-bank/research/message-lifecycle.md:175]: The observation about the formula is technically right, but the note already states it, ~30 lines below this bullet:

> A consequence worth stating: the read side applies `lo = max(catchup_min_cursor, start_block)` **before** evaluating completion. The literal `catchup_max_cursor < catchup_min_cursor` is wrong for a row written before seeding existed (`min = 0`, `max = start_block - 1`), which it would report as unfinished forever.

The split is deliberate rather than an oversight: the bullet states the cursor-algebra *model*, and the read-side guard is called out separately as a consequence of it. `CatchupProgress::compute`'s docstring is organised the same way — the guard gets its own paragraph there and is marked load-bearing. Folding it into the bullet would duplicate the existing paragraph.

Worth noting this PR did not touch that formula either: the only change to these lines is the trailing sentence distinguishing scan completion from the endpoint's `catchup_complete`.

Not acting on this one.
